File: local_examples/rust-sync/dt-string.rs
// EXAMPLE: set_tutorial
#[cfg(test)]
mod strings_tests {
    // STEP_START import
    use redis::{Commands, ExistenceCheck};
    // STEP_END

    #[test]
    fn run() {
        let mut r = match redis::Client::open("redis://127.0.0.1") {
            Ok(client) => {
                match client.get_connection() {
                    Ok(conn) => conn,
                    Err(e) => {
                        println!("Failed to connect to Redis: {e}");
                        return;
                    }
                }
            },
            Err(e) => {
                println!("Failed to create Redis client: {e}");
                return;
            }
        };

        // STEP_START set_get
        if let Ok(res) = r.set("bike:1", "Deimos") {
            let res: String = res;
            println!("{res}");    // >>> OK
            // REMOVE_START
            assert_eq!(res, "OK");
            // REMOVE_END
        }

        match r.get("bike:1") {
            Ok(res) => {
                let res: String = res;
                println!("{res}");   // >>> Deimos
                // REMOVE_START
                assert_eq!(res, "Deimos");
                // REMOVE_END
            },
            Err(e) => {
                println!("Error getting bike:1: {e}");
                return;
            }
        };
        // STEP_END

        // STEP_START setnx_xx
        if let Ok(res) = r.set_options("bike:1", "bike", redis::SetOptions::default().conditional_set(ExistenceCheck::NX)) {
            let res: bool = res;
            println!("{res}");    // >>> false
            // REMOVE_START
            assert!(!res);
            // REMOVE_END
        }

        match r.get("bike:1") {
            Ok(res) => {
                let res: String = res;
                println!("{res}");   // >>> Deimos
                // REMOVE_START
                assert_eq!(res, "Deimos");
                // REMOVE_END
            },
            Err(e) => {
                println!("Error getting bike:1: {e}");
                return;
            }
        };

        if let Ok(res) = r.set_options("bike:1", "bike", redis::SetOptions::default().conditional_set(ExistenceCheck::XX)) {
            let res: String = res;
            println!("{res}");    // >>> OK
            // REMOVE_START
            assert_eq!(res, "OK");
            // REMOVE_END
        }
        
        match r.get("bike:1") {
            Ok(res) => {
                let res: String = res;
                println!("{res}");   // >>> bike
                // REMOVE_START
                assert_eq!(res, "bike");
                // REMOVE_END
            },
            Err(e) => {
                println!("Error getting bike:1: {e}");
                return;
            }
        };
        // STEP_END

        // STEP_START mset
        if let Ok(res) = r.mset(&[("bike:1", "Deimos"), ("bike:2", "Ares"), ("bike:3", "Vanth")]) {
            let res: String = res;
            println!("{res}");    // >>> OK
            // REMOVE_START
            assert_eq!(res, "OK");
            // REMOVE_END
        }

        match r.mget(&["bike:1", "bike:2", "bike:3"]) {
            Ok(res) => {
                let res: Vec<String> = res;
                println!("{res:?}");   // >>> ["Deimos", "Ares", "Vanth"]
                // REMOVE_START
                assert_eq!(res.len(), 3);
                assert_eq!(res[0], "Deimos");
                assert_eq!(res[1], "Ares");
                assert_eq!(res[2], "Vanth");
                // REMOVE_END
            },
            Err(e) => {
                println!("Error getting values: {e}");
                return;
            }
        };
        // STEP_END

        // STEP_START incr
        if let Ok(res) = r.set("total_crashes", 0) {
            let res: String = res;
            println!("{res}");    // >>> OK
            // REMOVE_START
            assert_eq!(res, "OK");
            // REMOVE_END
        }

        if let Ok(res) = r.incr("total_crashes", 1) {
            let res: i32 = res;
            println!("{res}");    // >>> 1
            // REMOVE_START
            assert_eq!(res, 1);
            // REMOVE_END
        }

        if let Ok(res) = r.incr("total_crashes", 10) {
            let res: i32 = res;
            println!("{res}");    // >>> 11
            // REMOVE_START
            assert_eq!(res, 11);
            // REMOVE_END
        }
        // STEP_END
    }
}

File: local_examples/ruby/dt_string.rb
# EXAMPLE: set_tutorial
# HIDE_START
require 'redis'

r = Redis.new
# HIDE_END

# STEP_START set_get
res1 = r.set('bike:1', 'Deimos')
puts res1 # OK

res2 = r.get('bike:1')
puts res2 # Deimos
# STEP_END

# REMOVE_START
raise "Expected 'OK'" unless res1 == 'OK'
raise "Expected 'Deimos'" unless res2 == 'Deimos'
# REMOVE_END

# STEP_START setnx_xx
res3 = r.set('bike:1', 'bike', nx: true)
puts res3 # false

puts r.get('bike:1') # Deimos

res4 = r.set('bike:1', 'bike', xx: true)
puts res4 # true
# STEP_END

# REMOVE_START
raise 'Expected SET NX to fail' unless res3 == false
raise 'Expected SET XX to succeed' unless res4 == true
raise "Expected 'bike'" unless r.get('bike:1') == 'bike'
# REMOVE_END

# STEP_START mset
res5 = r.mset('bike:1', 'Deimos', 'bike:2', 'Ares', 'bike:3', 'Vanth')
puts res5 # OK

res6 = r.mget('bike:1', 'bike:2', 'bike:3')
puts res6.inspect # ["Deimos", "Ares", "Vanth"]
# STEP_END

# REMOVE_START
raise "Expected 'OK'" unless res5 == 'OK'
raise 'Expected all bike names' unless res6 == ['Deimos', 'Ares', 'Vanth']
# REMOVE_END

# STEP_START incr
r.set('total_crashes', 0)

res7 = r.incr('total_crashes')
puts res7 # 1

res8 = r.incrby('total_crashes', 10)
puts res8 # 11
# STEP_END

# REMOVE_START
raise 'Expected 1' unless res7 == 1
raise 'Expected 11' unless res8 == 11
r.close
# REMOVE_END
